src/core/Akka.Remote/EndpointManager.cs

```diff
@@ -492,196 +543,248 @@ protected void Accepting(object message)
-                .With<Terminated>(terminated =>
-                {
-                    AcceptPendingReader(terminated.ActorRef);
-                    endpoints.UnregisterEndpoint(terminated.ActorRef);
-                    HandleStashedInbound(terminated.ActorRef);
-                })
-                .With<EndpointWriter.TookOver>(tookover => RemovePendingReader(tookover.Writer, tookover.ProtocolHandle))
-                .With<ReliableDeliverySupervisor.GotUid>(gotuid =>
-                {
-                    endpoints.RegisterWritableEndpointUid(Sender, gotuid.Uid);
-                    HandleStashedInbound(Sender);
-                })
-                .With<Prune>(prune => endpoints.Prune())
-                .With<ShutdownAndFlush>(shutdown =>
-                {
+                            createAndRegisterWritingEndpoint(quarantined.Uid).Tell(send);
+                    })
+                    .Default(msg => createAndRegisterWritingEndpoint(null).Tell(send));
+            });
+            Receive<InboundAssociation>(ia => HandleInboundAssociation(ia, false));
+            Receive<EndpointWriter.StoppedReading>(endpoint => AcceptPendingReader(endpoint.Writer));
+            Receive<Terminated>(terminated =>
+            {
+                AcceptPendingReader(terminated.ActorRef);
+                _endpoints.UnregisterEndpoint(terminated.ActorRef);
+                HandleStashedInbound(terminated.ActorRef, writerIsIdle: false);
+            });
+            Receive<EndpointWriter.TookOver>(tookover => RemovePendingReader(tookover.Writer, tookover.ProtocolHandle));
+            Receive<ReliableDeliverySupervisor.GotUid>(gotuid =>
+            {
+                _endpoints.RegisterWritableEndpointUid(Sender, gotuid.Uid);
+                HandleStashedInbound(Sender, writerIsIdle: false);
+            });
+            Receive<ReliableDeliverySupervisor.Idle>(idle =>
+            {
+                HandleStashedInbound(Sender, writerIsIdle: true);
+            });
+            Receive<Prune>(prune => _endpoints.Prune());
+            Receive<ShutdownAndFlush>(shutdown =>
+            {
                     //Shutdown all endpoints and signal to Sender when ready (and whether all endpoints were shutdown gracefully)
                     var sender = Sender;
 
                     // The construction of the Task for shutdownStatus has to happen after the flushStatus future has been finished
                     // so that endpoints are shut down before transports.
-                    var shutdownStatus = Task.WhenAll(endpoints.AllEndpoints.Select(
-                            x => x.GracefulStop(settings.FlushWait, EndpointWriter.FlushAndStop.Instance))).ContinueWith(
-                                result =>
+                    var shutdownStatus = Task.WhenAll(_endpoints.AllEndpoints.Select(
+                        x => x.GracefulStop(_settings.FlushWait, EndpointWriter.FlushAndStop.Instance))).ContinueWith(
```

Comment:
This task completes when all of the endpoints managed by this `EndpointManager` have been gracefully stopped.